### PR TITLE
Add options libxml2:zlib libxml2:lzma libxml2:iconv

### PR DESCRIPTION
**Pull request by puetzk**
Offer defaults that match the previous hard-coded behavior, but allow consumers the option to include or exclude these optional features (and their corresponding requirments on other libraries).

Much-delayed in actually getting this sent, but as previously discussed over at https://cpplang.slack.com/archives/C77T8CBFB/p1539883027000100

### CODE REVIEW

**Comment by solvingj** [comment]
This is a great PR, thank you very much!   There is one difficulty, which is that the package is already in Conan Center, and the PR has a lot of potential to be a breaking change. 

The good news is that libxml 2.9.9 official release is imminent.  As soon as it is released, we will create `testing/2.9.9` branch, and you can re-submit this PR against that branch. 

Let me know what you think @puetzk .

Of note, this is why the Conan revisions feature is so important, it should enable us to make "breaking improvements" to existing packages like this moving forward, without actually breaking any users.

**Comment by puetzk** [comment]
It shouldn't be a breaking change for consumers unless you consider "alters package_id" to be breaking. The default_options match the previous hardcoded behavior, they just show up in info.options now (and that changes the hashes).  I suppose one could put logic in package_id to delete options that match defaults to keep those binarie ids unchanged, though that seems more confusing than helpful.

In any case, this is just (most of) what we're already using internally on a patched internal channel, I just wanted to take @SSE4 up on his offer that some of our changes were suitable to merge upstream. And finally got the approval through https://github.com/JDContributions to do so. If they're acceptable in principle but now is a bad time to take it, that's not a huge issue (though it would be nice to get them into a "testing" branch as soon as possible to avoid conflicts accumulating.

And I suppose if there's a 2.9.9 coming soon I should apparently get to work on submitting our other patches upstream to libxml2 ASAP :-)

**Comment by theirix** [comment]
I think we should wait for 2.9.9 (or 3.0.0). Also, we should look ahead for additional options for the Conan package.

**Comment by SSE4** [comment]
I personally don't think adding options is a breaking change (e.g. we have recently added new option `minizip` to the zlib package). but if you think it should wait, and release is coming soon, then let's wait.

File: conanfile.py [hunk 1]
```diff
@@ -91,20 +105,26 @@ def _build_with_configure(self):
...
                     tools.replace_in_file("configure", r"-install_name \$rpath/", "-install_name ")
-                configure_args = ['--with-python=no', '--without-lzma', '--prefix=%s' % full_install_subfolder]
+                configure_args = ['--with-python=no', '--prefix=%s' % full_install_subfolder]
                 if env_build.fpic:
                     configure_args.extend(['--with-pic'])
                 if self.options.shared:
                     configure_args.extend(['--enable-shared', '--disable-static'])
                 else:
                     configure_args.extend(['--enable-static', '--disable-shared'])
+                if not self.options.zlib:
```
Comment (SSE4) [inline conanfile.py]: these 3 extra conditions aren't needed, just use `configure_args.extend`

**Review by SSE4 — CHANGES_REQUESTED** [review]

**Comment by solvingj** [comment]
actually, it's not options that I consider breaking, although indeed even that is a gray area.  There's quite a bit of imperative code with conditions and string manipulation in there, which call into external scripts, the implications of changing which are hard to validate and predict.  With a package that is likely to be relied upon by so many downstreams, we should treat conan center package versions on the stable channel as immutable, at least until revisions is in place, assuming it does indeed protect cases like this.

File: conanfile.py (hunk 1 above)
Comment (puetzk) [reply conanfile.py]: Good point, changed

**Comment by puetzk** [comment]
>  There's quite a bit of imperative code

It's mostly indentation to move the MSVC build rules inside a tools.vcvars(...) context manager, if you ignore the leading spaces in the diff there's a lot less (but some, yes). Just in case you saw the big deleted/added block and didn't notice that.

**Comment by puetzk** [comment]
> I think we should wait for 2.9.9 (or 3.0.0)

I can't see waiting for 3.0.0, that's almost just "reject this PR forever" since I'm not aware of any upstream plans for a breaking major version.

In any case, there changes have nothing to do with the upstream version (it has supported these --with/--without options for years) except that having it change in any way conveniently covers for the current lack of "conan revisions" to label changes to the conan packaging of the same upstream version.

**Comment by solvingj** [comment]
2.9.9 is fine for sure.

**Comment by puetzk** [comment]
I made @SSE4's requested change, but I don't see how the appveyor failure could be related to my changes? 


> [zlib/1.2.11@conan/stable:f9c716b78ebb169925e958b727bbe35baf7c9fac - Missing](https://ci.appveyor.com/project/BinCrafters/conan-libxml2/builds/21358279/job/n5dsf72ir3mmiwb4#L1360)
> ...
> Can't find a 'zlib/1.2.11@conan/stable' package for the specified settings, options and dependencies:
> - Settings: arch=x86, build_type=Debug, compiler=Visual Studio, compiler.runtime=MDd, compiler.version=12, os=Windows
> - Options: minizip=False, shared=False

The exact same package(s) were requested for yesterday's build, and that were found...

> [zlib/1.2.11@conan/stable:8acc013b124da976880660b17a7d8021e2f49ff4 - Download](https://ci.appveyor.com/project/BinCrafters/conan-libxml2/builds/21345187/job/45evg2eun36xq567?fullLog=true#L348)
> ...
> [zlib/1.2.11@conan/stable:f9c716b78ebb169925e958b727bbe35baf7c9fac - Cache](https://ci.appveyor.com/project/BinCrafters/conan-libxml2/builds/21345187/job/45evg2eun36xq567?fullLog=true#L2368)

But interestingly, this points out that libxml2 and its test_package are picking different configurations of zlib (compiler.runtime is MTd in libxml2 and MDd in test_package). I didn't cause that, but was it intentional?

File: conanfile.py [hunk 2]
```diff
@@ -91,20 +105,23 @@ def _build_with_configure(self):
...
                     tools.replace_in_file("configure", r"-install_name \$rpath/", "-install_name ")
-                configure_args = ['--with-python=no', '--without-lzma', '--prefix=%s' % full_install_subfolder]
+                configure_args = ['--with-python=no', '--prefix=%s' % full_install_subfolder]
                 if env_build.fpic:
                     configure_args.extend(['--with-pic'])
                 if self.options.shared:
                     configure_args.extend(['--enable-shared', '--disable-static'])
                 else:
                     configure_args.extend(['--enable-static', '--disable-shared'])
+                configure_args.extend(['--with-zlib' if self.options.zlib else '--without-zlib'])
```
Comment (uilianries) [inline conanfile.py:115]:
```suggestion
                configure_args.extend(['--with-zlib={}'.format(self.deps_cpp_info["zlib"].libpaths[0]) if self.options.zlib else '--without-zlib'])
```

**Pull request merged**